File: iOS/TUILiveKit/Sources/Component/NetWorkInfo/Manager/NetWorkInfoService.swift
//
//  NetWorkInfoService.swift
//  Pods
//
//  Created by ssc on 2025/5/9.
//

import Foundation
import Network
import RTCRoomEngine
#if canImport(TXLiteAVSDK_TRTC)
import TXLiteAVSDK_TRTC
#elseif canImport(TXLiteAVSDK_Professional)
import TXLiteAVSDK_Professional
#endif

protocol NetWorkInfoServiceDelegate {
    func addTRTCObserver(_ observer: TRTCCloudDelegate)
    func removeTRTCObserver(_ observer: TRTCCloudDelegate)
}

class NetWorkInfoService {

    private let trtcCloud: TRTCCloud
    private let roomEngine: TUIRoomEngine = TUIRoomEngine.sharedInstance()

    init(trtcCloud: TRTCCloud) {
        self.trtcCloud = trtcCloud
    }

    func setAudioCaptureVolume(volume : Int) {
        trtcCloud.setAudioCaptureVolume(volume)
    }

    func getSelfUserId() -> String {
        return TUIRoomEngine.getSelfInfo().userId
    }

    func getVolueme() -> Int {
        return trtcCloud.getAudioCaptureVolume()
    }

    func setVideoResolution(resolution: TRTCVideoResolution) {
        let params = TRTCVideoEncParam()
        params.videoResolution = resolution
        trtcCloud.setVideoEncoderParam(params)
    }

    func updateAudioQuality(quality: TUIAudioQuality) {
        roomEngine.updateAudioQuality(quality)
    }
}

extension NetWorkInfoService: NetWorkInfoServiceDelegate {
    func addTRTCObserver(_ observer: any TRTCCloudDelegate) {
        trtcCloud.addDelegate(observer)
    }

    func addRoomEngineObserver(_ observer: any TUIRoomObserver) {
        roomEngine.addObserver(observer)
    }

    func removeTRTCObserver(_ observer: any TRTCCloudDelegate) {
        trtcCloud.removeDelegate(observer)
    }

    func removeRoomEngineObserver(_ observer: any TUIRoomObserver) {
        roomEngine.removeObserver(observer)
    }
}

File: Flutter/livekit/lib/component/network_info/manager/network_info_service.dart
import 'package:tencent_live_uikit/tencent_live_uikit.dart';
import 'package:tencent_rtc_sdk/trtc_cloud.dart';
import 'package:tencent_rtc_sdk/trtc_cloud_def.dart';
import 'package:tencent_rtc_sdk/trtc_cloud_listener.dart';

class NetworkInfoService {
  final TUIRoomEngine roomEngine = TUIRoomEngine.sharedInstance();
  TRTCCloud? trtcCloud;

  void addTRTCObserver(TRTCCloudListener observer) async {
    final trtcCloud = await _getTRTCCloud();
    trtcCloud.registerListener(observer);
  }

  void removeTRTCObserver(TRTCCloudListener observer) async {
    final trtcCloud = await _getTRTCCloud();
    trtcCloud.unRegisterListener(observer);
  }

  void addRoomEngineObserver(TUIRoomObserver observer) {
    roomEngine.addObserver(observer);
  }

  void removeRoomEngineObserver(TUIRoomObserver observer) {
    roomEngine.removeObserver(observer);
  }

  void setAudioCaptureVolume(int volume) async {
    final trtcCloud = await _getTRTCCloud();
    trtcCloud.setAudioCaptureVolume(volume);
  }

  String getSelfUserId() {
    return TUIRoomEngine.getSelfInfo().userId;
  }

  Future<int> getVolume() async {
    final trtcCloud = await _getTRTCCloud();
    return trtcCloud.getAudioCaptureVolume();
  }

  void setVideoResolution(TRTCVideoResolution resolution) async {
    final params = TRTCVideoEncParam();
    params.videoResolution = resolution;
    final trtcCloud = await _getTRTCCloud();
    trtcCloud.setVideoEncoderParam(params);
  }

  void updateAudioQuality(TUIAudioQuality quality) {
    roomEngine.updateAudioQuality(quality);
  }
}

extension on NetworkInfoService {
  Future<TRTCCloud> _getTRTCCloud() async {
    trtcCloud ??= await TRTCCloud.sharedInstance();
    return trtcCloud!;
  }
}
